### ISSUE TITLE
Image conversion timing out on Hyper-V

### ISSUE THREAD

**Issue by robogeek** [canonical/multipass#1522]
**Describe the bug**

While exploring timeouts on launching Multipass instances on Windows (see https://github.com/canonical/multipass/issues/1512) I experimented further, and narrowed down one clear bug.  That launching a 20.04 instance results in a timeout, while launching an 18.04 instance does not.

The general form of the timeout is this:

```
PS C:\Users\heavy> multipass launch
launch failed: The following errors occurred:
unassuming-tailorbird: timed out waiting for response
```

My previous filed issue went over this as a general problem.  After I got my computer to a working state, where it readily launched and managed instances.  I experimented some more, and found the following pattern.

* `multipass launch` (defaults to 18.04) launches successfully
* `multipass launch focal` (to launch 20.04) fails with the timeout 

I could launch 18.04 (success) then try 20.04 (fail) followed by 18.04 (success) and followed by 20.04 (fail).  Meaning that once a launch failed it did not ruin the system requiring a reboot.

**To Reproduce**

* `multipass launch` (defaults to 18.04) launches successfully
* `multipass launch focal` (to launch 20.04) fails with the timeout 

**Expected behavior**

It should launch 20.04 just as readily as it does 18.04

**Logs**

**Additional info**
 - OS: Windows 10 Pro Insider Preview - Version 2004 - Installed on 4/27/2020 - OS Build 19612.1005 - Windows Feature Experience Pack 120.8400.0.0
- 1.2.1
- WSL2 was not enabled at the time

**Additional context**

This report is further experimentation after I filed https://github.com/canonical/multipass/issues/1512

I've written a loooooong blog post describing a bunch of details.  At the bottom is a diary of the experimentation I did.  The above report comes from one small portion of that diary.  See: https://techsparx.com/linux/multipass/windows/troubleshoot.html

There was no corruption of `hosts.ics` during the experiment of 18.04 versus 20.04.  There are plenty of other ways that `hosts.ics` can become corrupt, just not this particular instance.

For the instances that timed out - `multipass list` showed the status as Running, but no IP address was assigned.  An attempt to stop these instances (so I could delete/purge them) would hang, with the longest recorded hang being 2 hours.  But as soon as I rebooted, `multipass list` showed them in Suspended state and I could stop/delete/purge these instances.

**Comment by Saviq** [comment]
Hi @robogeek, I've been trying to reproduce your issues and could not - whether I used Hyper-V or VirtualBox, `focal` launched fine on a fully updated stable Windows.

I'm installing insiders' preview now to try as well.

When you open the graphical console, can you see the instance booting ok?

**Comment by robogeek** [comment]
Yes.. it starts showing in the Running state as soon as the Multipass progress output shows it in the Starting state.  It stays in the Running state after Multipass prints the timeout message.  In the bottom pane of Hyper-V, the Networking tab shows it connected to the Default Switch but no IP address is assigned.

```
PS C:\Users\heavy> multipass list
Name                    State                IPv4             Image
focal                      Running           UNKNOWN          Ubuntu 20.04 LTS
svc-notes               Stopped           --               Ubuntu 18.04 LTS
svc-userauth          Stopped           --               Ubuntu 18.04 LTS

PS C:\Users\heavy> multipass info --all
info failed: ssh connection failed: 'Failed to resolve hostname focal.mshome.net (No such host is known. )'

PS C:\Users\heavy> multipass version
multipass  1.2.1+win
multipassd 1.2.1+win
```

FWIW, `hosts.ics` has some corruption:

```
# Copyright (c) 1993-2001 Microsoft Corp.
#
# This file has been automatically generated for use by Microsoft Internet
# Connection Sharing. It contains the mappings of IP addresses to host names
# for the home network. Please do not make changes to the HOSTS.ICS file.
# Any changes may result in a loss of connectivity between machines on the
# local network.
#

172.23.80.1 HEAVY.mshome.net # 2025 5 0 4 15 24 20 791
172.23.89.142 svc-notes.mshome.net # 2020 5 2 12 1 17 35 805
172.23.83.119 svc-userauth.mshome.net # 2020 5 2 12 1 15 38 250
50
```

Using the Connect button -- gave me a screen talking about _grub rescue mode_

I tried `multipass stop focal` and that simply said Stopping with the spinning indicator for a long time, with no changes in the Hyper-V console.  After about 3-4 minutes I gave up on that (CTRL-C'd it).

Then tried `multipass launch` to try and launch an 18.04 instance.  It too is waiting a long time without printing anything.

While that is underway, in another window I tried `multipass list` and that too is taking a long time to respond.

While that is underway, in Hyper-V I selected `svc-userauth` and clicked the Start button.  Hyper-V shows it starting up, then I clicked the Connect button and am seeing normal Linux startup messages.  It eventually gave me a login prompt.  But since I don't know the password I can't go further.

In other words, it seems the Multipass daemon is locked up on something or other....?  In Task Manager, the processes with _multipass_ in their name are showing in Running state.

**Comment by Saviq** [comment]
> but no IP address is assigned.

This only works if the running instance knows how to communicate the IP to Hyper-V, which I don't think Ubuntu knows how…

> Using the Connect button -- gave me a screen talking about _grub rescue mode_

This sounds interesting. Can you give a screenshot of this state, please?

> In other words, it seems the Multipass daemon is locked up on something or other....? In Task Manager, the processes with _multipass_ in their name are showing in Running state.

That we can confirm, it's a bit thick-headed when things don't go as expected.

But please let's focus on the `focal` instance. You said above it never booted. Please just start it in Hyper-V and let us know if you see it booting normally, or as you said - stuck on GRUB - and what exactly is printed.

**Comment by robogeek** [comment]
In the past what was printed in the Connect screen was:

```
Error: invalid arch-independend ELF magic.
Entering rescue mode...
grub rescue>
```

Currently the Connect screen shows black, with a blinking cursor in a place roughly close to what would be the end of that text.

![P1000524](https://user-images.githubusercontent.com/358930/81093360-d9f31500-8eb6-11ea-97a0-0c2d79fb3322.JPG)

On a lark, I hit RETURN and it got to this:

![P1000526](https://user-images.githubusercontent.com/358930/81093430-f5f6b680-8eb6-11ea-872a-03491693b8f5.JPG)

To get to this state required a couple reboots.  The first was after using Notepad to clean up the corruption in `hosts.ics`, and the second was after cleaning up corruption in `hosts.ics` on that reboot.  Then there were at least two further reboots to be able to observe the startup process of Hyper-V.

On booting the system, `focal` is shown in Hyper-V in the Saved state.  Starting the Connect window, it has a message about `focal` being turned off and I can click the Start button to start it.  In `multipass list` it is shown as Suspended.

Click _Start_ in the Connect window.  The message in the Connect window changes to _Restoring_ then to _Connecting_ then to the first of the above images.

**Comment by Saviq** [comment]
@robogeek OK so the instance never boots, no wonder it's not getting IPs.

Can you try a fresh `multipass launch focal` and monitor it in Hyper-V?

**Comment by robogeek** [comment]
It took a couple reboots before I could get it to the state where the old `focal` instance could be deleted.

Ran: `multipass launch --name focal2 focal` to start a fresh instance under a new name just in case there was something left over from the old name.

I had Hyper-V already launched.  It took a moment or so before Hyper-V showed the instance.  As quickly as I could I got the Connection window launched but that wasn't until Multipass showed it was Starting the instance and Hyper-V already showed it in Running state.

The first message I found after getting the Connection window to connect was this:

![P1000531](https://user-images.githubusercontent.com/358930/81100402-54c12d80-8ec1-11ea-8cfc-279a783ec9f8.JPG)

This capture was made before Multipass had timed out - meaning Multipass still said Starting and was spinning the spinner.  Eventually it gave this:

```
PS C:\Users\heavy> multipass launch --name focal2 focal
launch failed: The following errors occurred:
focal2: timed out waiting for response
```

I have no clue what to do in Grub Rescue Mode and there's no help command.

**Comment by Saviq** [comment]
Oh wow, @robogeek is your Windows 32-bit by any chance?

**Comment by robogeek** [comment]
No ...  It's an HP EliteBook 2650p w/ Core i7 @ 2.8 GHz w/ 16 GB of RAM.  It is installed as 64 bit Windows 10 Pro.  The biggest issue is the drive is a 128GB SSD and it's nearly full.

**Comment by Saviq** [comment]
@robogeek the only reason I can think of this happening is the image got corrupted somehow. Can you try `multipass launch daily:focal` to pull a newer image? If that works, we can guide you how to remove the invalid release image.

**Comment by Saviq** [comment]
@robogeek I believe I understand what happened - we had too short a timeout for image conversion, which must have resulted in a corrupted image. I reproduced that inside a VM and we have a fix coming.

**Comment by robogeek** [comment]
Okay.  My laptop isn't the fastest around.  Thank you.  Let me know whether there's something I can do to help test.

**Comment by Saviq** [comment]
You can! Here's a preview of that fix: [multipass-1.3.0-dev.235+g0e3d34d9.win-win64.exe](https://multipass-ci.s3.amazonaws.com/ppr241/multipass-1.3.0-dev.235%2Bg0e3d34d9.win-win64.exe).

If you can, answer "Yes" when the uninstaller asks you whether to remove all data. That'd be simplest to get to a clean slate. If not, let us know and we'll guide you through what to do.

**Comment by robogeek** [comment]
I uninstalled 1.2.1 and as you said asked it to remove all data.  

```
PS C:\Users\heavy> multipass launch --name focal3 focal
launch failed: The following errors occurred:
focal3: timed out waiting for response
PS C:\Users\heavy> multipass list
Name                    State             IPv4             Image
focal3                  Running           172.21.73.38     Ubuntu 20.04 LTS
PS C:\Users\heavy> multipass shell focal3
Welcome to Ubuntu 20.04 LTS (GNU/Linux 5.4.0-29-generic x86_64)

 * Documentation:  https://help.ubuntu.com
 * Management:     https://landscape.canonical.com
 * Support:        https://ubuntu.com/advantage

 System information disabled due to load higher than 1.0

0 updates can be installed immediately.
0 of these updates are security updates.



The programs included with the Ubuntu system are free software;
the exact distribution terms for each program are described in the
individual files in /usr/share/doc/*/copyright.

Ubuntu comes with ABSOLUTELY NO WARRANTY, to the extent permitted by
applicable law.

To run a command as administrator (user "root"), use "sudo <command>".
See "man sudo_root" for details.

ubuntu@focal3:~$
```

So..  It got the timeout, but went ahead and assigned an IP address anyway.  I had attached a Connect window during the startup, and saw the normal Linux bootup messages.  Then after the timeout occurred, the Connect window switched to the Login prompt.  After seeing the login prompt I ran the `multipass list` shown above.

I saw the same behavior yesterday while launching an 18.04 instance.  That it went through starting up the instance, and that there was a timeout message, but after the timeout the instance was useful anyway.

I tried using the `daily:focal` image as you suggested a couple comments back.  But it gave me a message about not having access to the `daily`  repository.  But because it was a fresh clean slate install the `focal` image had to download anyway.

**Comment by Saviq** [comment]
Yes, we have a 5 minute timeout for the instance to come up, but if it does, later, it will still work.

So yeah, what you're getting now is expected - and things will work. We have some work to do on the timeout side so it's a better experience. Most importantly, we now know everything that was happening. Thanks for helping out!
